src/tiling.rs

```diff
@@ -0,0 +1,600 @@
+//! For tiling multiple displays together in various grid arrangements
+//! They have to be tiles together in some specific supported grid layouts.
+//! Currently supported layouts:
+//! - [`ChainTopRightDown`]
+//!
+//! To write to those panels the [`TiledFrameBuffer`] can be used.
+//! A usage example can be found at that structs documentation.
+
+use core::{convert::Infallible, marker::PhantomData};
+
+use crate::{Color, FrameBuffer, FrameBufferUser, WordSize};
+#[cfg(not(feature = "esp-hal-dma"))]
+use embedded_dma::ReadBuffer;
+use embedded_graphics::prelude::{PixelColor, Point};
+#[cfg(feature = "esp-hal-dma")]
+use esp_hal::dma::ReadBuffer;
+
+/// Computes the number of columns needed if the displays are bing tiled together.
+/// # Arguments
+///
+/// * `cols` - Number of columns per panel
+/// * `num_panels_wide` - Number of panels tiled horizontally
+/// * `num_panels_high` - Number of panels tiled vertically
+///
+/// # Returns
+///
+/// Number of columns needed internally for `DmaFrameBuffer`
+#[must_use]
+pub const fn compute_tiled_cols(
+    cols: usize,
+    num_panels_wide: usize,
+    num_panels_high: usize,
+) -> usize {
+    cols * num_panels_wide * num_panels_high
+}
+
+/// Trait for pixel re-mappers
+///
+/// Implementors of this trait will remap x,y coordinates from a
+/// virtual panel to the actual framebuffer used to drive the panels
+///
+/// # Type Parameters
+///
+/// * `PANEL_ROWS` - Number of rows in a single panel
+/// * `PANEL_COLS` - Number of columns in a single panel
+/// * `TILE_ROWS` - Number of panels stacked vertically
+/// * `TILE_COLS` - Number of panels stacked horizontally
+pub trait PixelRemapper {
+    /// Number of rows in the virtual panel
+    const VIRT_ROWS: usize;
+    /// Number of columns in the virtual panel
+    const VIRT_COLS: usize;
+    /// Number of rows in the actual framebuffer
+    const FB_ROWS: usize;
+    /// Number of columns in the actual framebuffer
+    const FB_COLS: usize;
+
+    /// Remap a virtual pixel to a framebuffer pixel
+    #[inline]
+    fn remap<C: PixelColor>(mut pixel: embedded_graphics::Pixel<C>) -> embedded_graphics::Pixel<C> {
+        pixel.0 = Self::remap_point(pixel.0);
+        pixel
+    }
+
+    /// Remap a virtual point to a framebuffer point
+    #[inline]
+    #[must_use]
+    fn remap_point(mut point: Point) -> Point {
+        let (re_x, re_y) = Self::remap_xy(point.x as usize, point.y as usize);
```

Comment:
If `x` or `y` is `< 0` then don't bother remapping since its off the screen and will be skipped during rendering and would generate garbage if remapped.